src/threads-version/big_int64.cpp

```
@@ -0,0 +1,767 @@
+  res.isNegative = a.isNegative != b.isNegative;
+  return res;
+}
+void BigInt64::rightShift(uint64_t n) { //--->
+  if (n >= bitsCount()) {
+    zero();
+    return;
+  }
+  uint64_t shiftEachLimb = n % BITS_IN_LIMB;
+  uint64_t dropLimbs = n / BITS_IN_LIMB;
+  if (shiftEachLimb != 0) {
+    for (int i = 0; i < limbsCount() - 1 - dropLimbs; i++) // lsb...msb
+      limbs[i] = limbs[i + dropLimbs] >> shiftEachLimb |
+                 limbs[i + dropLimbs + 1] << (BITS_IN_LIMB - shiftEachLimb);
+    limbs[limbsCount() - 1 - dropLimbs] >>= shiftEachLimb;
+    limbs.resize(limbsCount() - dropLimbs);
+    if (limbs.back() == 0 && !isZero())
+      limbs.pop_back();
+  } else
+    limbs.erase(limbs.begin(), limbs.begin() + dropLimbs);
+}
+
+void BigInt64::leftShift(uint64_t n) { //<---
+  if (n >= bitsCount()) {
+    zero();
+    return;
+  }
+  uint64_t shiftEachLimb = n % BITS_IN_LIMB;
+  uint64_t dropLimbs = n / BITS_IN_LIMB;
+  if (shiftEachLimb != 0) {
+    for (int i = limbsCount() - 1; i > dropLimbs; i--) { // msb-lst
+      limbs[i] = limbs[i - dropLimbs] << shiftEachLimb |
+                 limbs[i - dropLimbs - 1] >> (BITS_IN_LIMB - shiftEachLimb);
+    }
+    limbs[0] <<= shiftEachLimb;
+  } else {
+    for (int i = limbsCount() - 1; i >= dropLimbs; i--)
+      limbs[i] = limbs[i - dropLimbs];
+    for (int i = dropLimbs - 1; i >= 0; i--)
+      limbs[i] = 0;
+  }
+
+  removeLeadingZeros();
+}
+BigInt64 extendedGcd(BigInt64 a, BigInt64 b, BigInt64 &x, BigInt64 &y) {
+  x = 1;
+  y = 0;
+  BigInt64 x1 = 0, y1 = 1;
+  BigInt64 q, temp;
+
+  while (b != 0) {
+
+    q = a / b;
+    temp = b;
+    b = a % b;
+    a = temp;
+
+    temp = x1;
+
+    x1 = x - q * x1;
+    x = temp;
+
+    temp = y1;
+    y1 = y - q * y1;
+    y = temp;
+  }
+
+  return a;
+}
+BigInt64 modularInverse(BigInt64 a, BigInt64 b) {
+  BigInt64 x, y;
+  BigInt64 gcdResult = extendedGcd(a, b, x, y);
+  if (gcdResult != 1)
+    throw std::invalid_argument("Modular inverse does not exist.");
```

Should return an error code?

---

wrapped with try catch